CMakeLists.txt

```diff
@@ -0,0 +1,112 @@
+)
+set(CMAKE_VS_JUST_MY_CODE_DEBUGGING ON)
+
+# Collect baseline recommended warning flags.
+if(MSVC)
+    string(REGEX REPLACE "[-/]W[1-4]" "" CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}")
+    set(MM_REQUIRED_WARN_FLAGS ${MM_REQUIRED_WARN_FLAGS} /W4 /permissive-)
+else()
+    set(MM_REQUIRED_WARN_FLAGS -Wall -Wextra -Wpedantic -Wnon-virtual-dtor)
+endif()
+
+if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+    # GCC `missing-field-initializers` produces nonsense warnings about
+    # `<struct>::<anonymous>`.
+    set(MM_REQUIRED_WARN_FLAGS
+        ${MM_REQUIRED_WARN_FLAGS}
+        -Wno-missing-field-initializers
+    )
+endif()
+
+# See: https://github.com/cpp-best-practices/cppbestpractices.
+option(MM_USE_EXTRA_WARNINGS
+    "Enable more warnings recommended by C++ Best Practices." TRUE
+)
+if(MM_USE_EXTRA_WARNINGS)
+    if(MSVC)
+        set(MM_EXTRA_WARN_FLAGS
+            /w14242 # possible loss of data: conversion from `type1` to `type2`
+            /w14254 # possible loss of data: conversion from `type1:field_bits`
+                    # to `type2:field_bits`
+            /w14263 # member function does not override any base class virtual
+                    # member function
+            /w14265 # class has virtual functions, but destructor is not
+                    # virtual; instances of this class may not be destructed
+                    # correctly
+            /w14287 # unsigned/negative constant mismatch
+            /we4289 # loop control variable declared in the for-loop is used
+                    # outside the for-loop scope
+            /w14296 # expression is always [true | false]
+            /w14311 # pointer truncation from `type1` to `type2`
+            /w14545 # expression before comma evaluates to a function which is
+                    # missing an argument list
+            /w14546 # function call before comma missing argument list
+            /w14547 # operator before comma has no effect; expected operator
+                    # with side-effect
+            /w14549 # operator before comma has no effect; did you intend
+                    # 'operator'?
+            /w14555 # expression has no effect; expected expression with
+                    # side-effect
+            /w14619 # pragma warning: there is no warning number `number`
+            /w14640 # thread un-safe static member initialization
+            /w14826 # conversion from 'type1' to 'type_2' is sign-extended which
+                    # may cause unexpected runtime behavior
+            /w14928 # illegal copy-initialization; more than one user-defined
+                    # conversion has been implicitly applied
+        )
+    else()
+        set(MM_EXTRA_WARN_FLAGS
+            -Wcast-align
+            -Wunused
+            -Woverloaded-virtual
+            -Wconversion
+            -Wsign-conversion
+            -Wnull-dereference
+            -Wmisleading-indentation
+            -Wformat=2
+        )
+    endif()
```

Comment:
`-Wall -Wpedantic -Wextra`?

Also, what's `-flto=thin`? I see it in your Messenger name lol

Reply:
Wall et al. is [here](https://github.com/thedavidchu/merry_mem/blob/89443abe4baf0352ca24ff32939666d382a2e29e/CMakeLists.txt#L36).

flto=thin (as opposed to just flto) is IPO/whole program opt but less resource intensive and can be done incrementally.